File: frontend/src/api/roi.ts
import { apiPost } from './client'

export interface ROIRequest {
  purchase_price: number
  loan_amount: number
  annual_rate_pct?: number
  tenure_years?: number
  hold_years?: number
  monthly_rental_sgd?: number
  annual_appreciation_pct?: number
  annual_expenses_pct?: number
}

export interface YearlyResult {
  year: number
  property_value: number
  rental_income: number
  mortgage_payment: number
  annual_expenses: number
  net_cashflow: number
  cumulative_cashflow: number
  loan_balance: number
}

export interface ScenarioResult {
  label: string
  total_return_pct: number
  irr_pct: number
  net_gain: number
  exit_value: number
}

export interface ROIResponse {
  down_payment: number
  yearly: YearlyResult[]
  exit_value: number
  total_rental_income: number
  total_mortgage_paid: number
  total_expenses: number
  net_gain: number
  total_return_pct: number
  irr_pct: number
  scenarios: ScenarioResult[]
}

export function calculateROI(req: ROIRequest): Promise<ROIResponse> {
  return apiPost('/api/roi/calculate', req)
}

File: backend/app/schemas/roi.py
from pydantic import BaseModel, Field


class ROIRequest(BaseModel):
    purchase_price: float = Field(gt=0)
    loan_amount: float = Field(ge=0)
    annual_rate_pct: float = Field(default=3.5, ge=0.0, le=20.0)
    tenure_years: int = Field(default=25, ge=1, le=35)
    hold_years: int = Field(default=5, ge=1, le=35)
    monthly_rental_sgd: float = Field(default=0.0, ge=0.0)
    annual_appreciation_pct: float = Field(default=3.0, ge=-20.0, le=30.0)
    annual_expenses_pct: float = Field(default=1.0, ge=0.0, le=10.0)


class YearlyResult(BaseModel):
    year: int
    property_value: float
    rental_income: float
    mortgage_payment: float
    annual_expenses: float
    net_cashflow: float
    cumulative_cashflow: float
    loan_balance: float


class ScenarioResult(BaseModel):
    label: str
    total_return_pct: float
    irr_pct: float
    net_gain: float
    exit_value: float


class ROIResponse(BaseModel):
    down_payment: float
    yearly: list[YearlyResult]
    exit_value: float
    total_rental_income: float
    total_mortgage_paid: float
    total_expenses: float
    net_gain: float
    total_return_pct: float
    irr_pct: float
    scenarios: list[ScenarioResult]
